### ISSUE TITLE
neuronx-cc compiler crash `Invalid access of 8 partitions starting at partition 48`

### ISSUE THREAD

**Issue by xanderdunn** [aws-neuron/aws-neuron-sdk#695]
Compiling with
```
neuronx-cc compile ~/rust_hlo_run_rank_75.pb --framework XLA --target trn1 --model-type transformer --output /tmp/run_rank_75.neff
```

Here is a graph that compiles successfully: [rust_hlo_run_rank_76.pb.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/11803909/rust_hlo_run_rank_76.pb.zip)
Here is a graph that fails compilation: [rust_hlo_run_rank_75.pb.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/11803907/rust_hlo_run_rank_75.pb.zip)

Compiler error:
```
$ neuronx-cc compile ~/rust_hlo_run_rank_75.pb --framework XLA --target trn1 --model-type transformer --output /tmp/run_rank_75.neff
2023-06-20T16:48:59Z WARNING 22697 [StaticProfiler]: matmul-based transposes inserted by penguin takes up 87.90 percent of all matmul computation
2023-06-20T16:49:00Z ERROR 22697 [WalrusDriver]: An exception was thrown:
--------------------------------------------------------------------------------
 0# __cxa_throw in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libwalrus.so
 1# 0x00007F5362C1E3FB in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libBIRVerifier.so
 2# birverifier::checkLegalPartitionAccess(bir::Instruction const&, birverifier::InstIO, unsigned int, bir::AccessPattern const&) in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libBIRVerifier.so
 3# birverifier::InstVisitor::visitInstruction(bir::Instruction&) in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libBIRVerifier.so
 4# birverifier::InstVisitor::visitInstTensorCopy(bir::InstTensorCopy&) in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libBIRVerifier.so
 5# bir::IRVisitor<birverifier::InstVisitor, void>::visit(bir::Instruction&) in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libwalrus.so
 6# neuronxcc::walrus::Verifier::run(bir::Module&) in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libwalrus.so
 7# neuronxcc::walrus::WalrusPass::run(std::vector<std::unique_ptr<bir::Module, std::default_delete<bir::Module> >, std::allocator<std::unique_ptr<bir::Module, std::default_delete<bir::Module> > > >&) in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libwalrus.so
 8# 0x00007F527A889B1C in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libwalrus.so
 9# run_walrus_driver(int, char**) in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/../../../starfish/lib/libwalrus.so
10# 0x00007F5362C6009E in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/support/EmbeddedWalrusDriver.cpython-38-x86_64-linux-gnu.so
11# 0x00007F527D6FC973 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/WalrusDriver.cpython-38-x86_64-linux-gnu.so
12# _PyObject_MakeTpCall in /usr/bin/python3.8
13# _PyObject_FastCallDict in /usr/bin/python3.8
14# _PyObject_Call_Prepend in /usr/bin/python3.8
15# 0x00007F527D6E1C86 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/WalrusDriver.cpython-38-x86_64-linux-gnu.so
16# 0x00007F527D7104B9 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/WalrusDriver.cpython-38-x86_64-linux-gnu.so
17# _PyObject_MakeTpCall in /usr/bin/python3.8
18# _PyObject_FastCallDict in /usr/bin/python3.8
19# _PyObject_Call_Prepend in /usr/bin/python3.8
20# 0x00007F53A2968D06 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/Job.cpython-38-x86_64-linux-gnu.so
21# 0x00007F53A29801D3 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/Job.cpython-38-x86_64-linux-gnu.so
22# _PyObject_MakeTpCall in /usr/bin/python3.8
23# _PyObject_FastCallDict in /usr/bin/python3.8
24# _PyObject_Call_Prepend in /usr/bin/python3.8
25# 0x00007F53A2968D06 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/Job.cpython-38-x86_64-linux-gnu.so
26# 0x00007F53A297A463 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/Job.cpython-38-x86_64-linux-gnu.so
27# 0x00007F527D6E7215 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/jobs/WalrusDriver.cpython-38-x86_64-linux-gnu.so
28# _PyObject_MakeTpCall in /usr/bin/python3.8
29# _PyObject_FastCallDict in /usr/bin/python3.8
30# _PyObject_Call_Prepend in /usr/bin/python3.8
31# 0x00007F53A299AAB2 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/Pipeline.cpython-38-x86_64-linux-gnu.so
32# _PyObject_MakeTpCall in /usr/bin/python3.8
33# _PyObject_FastCallDict in /usr/bin/python3.8
34# _PyObject_Call_Prepend in /usr/bin/python3.8
35# 0x00007F53A2968D06 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/Job.cpython-38-x86_64-linux-gnu.so
36# 0x00007F53A29801D3 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/Job.cpython-38-x86_64-linux-gnu.so
37# _PyObject_MakeTpCall in /usr/bin/python3.8
38# _PyObject_FastCallDict in /usr/bin/python3.8
39# _PyObject_Call_Prepend in /usr/bin/python3.8
40# 0x00007F53A2968D06 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/Job.cpython-38-x86_64-linux-gnu.so
41# 0x00007F53A297A463 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/Job.cpython-38-x86_64-linux-gnu.so
42# _PyObject_MakeTpCall in /usr/bin/python3.8
43# _PyObject_FastCallDict in /usr/bin/python3.8
44# _PyObject_Call_Prepend in /usr/bin/python3.8
45# 0x00007F53A2428596 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/commands/CompileCommand.cpython-38-x86_64-linux-gnu.so
46# 0x00007F53A2457558 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/commands/CompileCommand.cpython-38-x86_64-linux-gnu.so
47# _PyObject_MakeTpCall in /usr/bin/python3.8
48# _PyObject_FastCallDict in /usr/bin/python3.8
49# _PyObject_Call_Prepend in /usr/bin/python3.8
50# 0x00007F53A242EA57 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/commands/CompileCommand.cpython-38-x86_64-linux-gnu.so
51# _PyObject_MakeTpCall in /usr/bin/python3.8
52# _PyObject_FastCallDict in /usr/bin/python3.8
53# _PyObject_Call_Prepend in /usr/bin/python3.8
54# 0x00007F53A2A6DC06 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/CommandDriver.cpython-38-x86_64-linux-gnu.so
55# 0x00007F53A2A8045A in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/CommandDriver.cpython-38-x86_64-linux-gnu.so
56# _PyObject_MakeTpCall in /usr/bin/python3.8
57# _PyObject_FastCallDict in /usr/bin/python3.8
58# _PyObject_Call_Prepend in /usr/bin/python3.8
59# 0x00007F53A2A6FB3B in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/CommandDriver.cpython-38-x86_64-linux-gnu.so
60# 0x00007F53A2A6FD12 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/CommandDriver.cpython-38-x86_64-linux-gnu.so
61# 0x00007F53A2A7CBD5 in /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/driver/CommandDriver.cpython-38-x86_64-linux-gnu.so
62# _PyObject_MakeTpCall in /usr/bin/python3.8
63# _PyEval_EvalFrameDefault in /usr/bin/python3.8
64# _PyEval_EvalCodeWithName in /usr/bin/python3.8
65# PyEval_EvalCode in /usr/bin/python3.8
66# 0x000000000067DBF1 in /usr/bin/python3.8
67# 0x000000000067DC6F in /usr/bin/python3.8
68# 0x000000000067DD11 in /usr/bin/python3.8
69# PyRun_SimpleFileExFlags in /usr/bin/python3.8
70# Py_RunMain in /usr/bin/python3.8
71# Py_BytesMain in /usr/bin/python3.8
72# __libc_start_main in /lib/x86_64-linux-gnu/libc.so.6
73# _start in /usr/bin/python3.8
--------------------------------------------------------------------------------
2023-06-20T16:49:00Z ERROR 22697 [WalrusDriver]: Walrus pass: birverifier failed!
2023-06-20T16:49:00Z ERROR 22697 [WalrusDriver]: Failure Reason: === BIR verification failed ===
Reason: Invalid access of 8 partitions starting at partition 88
Instruction: I-2926
Opcode: TensorCopy
Input index: 0
Argument AP:
Access Pattern: [[1024,8],[1024,1],[1,512]]
SymbolicAP
Memory Location: {mhlo_multiply_106_pftranspose_1533_set}@SB
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]: ***************************************************************
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:  An Internal Compiler Error has occurred
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]: ***************************************************************
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]: Error message:  [Errno 2] No such file or directory: 'sg00/def.json'
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]: Error class:    FileNotFoundError
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]: Error location: Unknown
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]: Command line:   /home/ubuntu/.local/bin/neuronx-cc compile /home/ubuntu/rust_hlo_run_rank_75.pb --framework XLA --target trn1 --model-type transformer --output /tmp/run_rank_75.neff
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]: Internal details:
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/CommandDriver.py", line 256, in neuronxcc.driver.CommandDriver.CommandDriver.run
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/commands/CompileCommand.py", line 1099, in neuronxcc.driver.commands.CompileCommand.CompileCommand.run
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/commands/CompileCommand.py", line 1050, in neuronxcc.driver.commands.CompileCommand.CompileCommand.runPipeline
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/commands/CompileCommand.py", line 1075, in neuronxcc.driver.commands.CompileCommand.CompileCommand.runPipeline
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/commands/CompileCommand.py", line 1079, in neuronxcc.driver.commands.CompileCommand.CompileCommand.runPipeline
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/Job.py", line 300, in neuronxcc.driver.Job.SingleInputJob.run
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/Job.py", line 326, in neuronxcc.driver.Job.SingleInputJob.runOnState
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/Pipeline.py", line 30, in neuronxcc.driver.Pipeline.Pipeline.runSingleInput
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/jobs/Kelper.py", line 335, in neuronxcc.driver.jobs.Kelper.Kelper.run
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:   File "neuronxcc/driver/jobs/Kelper.py", line 98, in neuronxcc.driver.jobs.Kelper.Kelper.genKelfFiles
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]:
2023-06-20T16:49:00Z ERROR 22697 [neuronx-cc]: Version information:
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]:   NeuronX Compiler version 2.7.0.40+f7c6cf2a3
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]:
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]:   Running on AMI ami-01257e71ecb2f431c
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]:   Running in region usw2-az4
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]:   Running on python 3.8.10
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]:
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]:   HWM version 2.7.0.3-0092b9d34
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]:   NumPy version 1.21.6
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]:
2023-06-20T16:49:01Z ERROR 22697 [neuronx-cc]: Artifacts stored in: /home/ubuntu/dev/Kholinar/xla/neuronxcc-j7uhyqy6
```

This is interesting because there are only three minor differences between the graphs:
Comparing ranks 75, left, (crash) and 76, right, (no crash):
<img width="1263" alt="Screenshot 2023-06-20 at 12 48 18" src="https://github.com/JasnahOrg/Kholinar/assets/1313618/b39fc2e8-71aa-4a6e-b195-099d05aa796f">
<img width="1263" alt="Screenshot 2023-06-20 at 12 48 22" src="https://github.com/JasnahOrg/Kholinar/assets/1313618/3737b54c-50e0-4a08-bd3b-4b04cf8416a0">
<img width="1267" alt="Screenshot 2023-06-20 at 12 48 26" src="https://github.com/JasnahOrg/Kholinar/assets/1313618/28d8eac7-e674-4c37-b14e-b8815a7d81b4">

For context, each graph here is 1 out of 96 ranks in a transformer with pipeline parallelism with these hyper-parameters:
```
{
  "d_context": 256,
  "d_model": 1024,
  "n_heads": 32,
  "n_layers": 3,
}
```
Across three trn1n.32xlarge instances, each with 32 ranks.

Version
```
$ neuronx-cc --version
NeuronX Compiler version 2.7.0.40+f7c6cf2a3

Running on AMI ami-01257e71ecb2f431c
Running in region usw2-az4
Running on python 3.8.10

HWM version 2.7.0.3-0092b9d34
NumPy version 1.21.6
```

**Comment by aws-donkrets** [comment]
Hi [xanderdunn](https://github.com/xanderdunn) - thx for reporting this issue.
We have reproduced it and are testing a fix. Will update this ticket with release availability when more info is known.

**Comment by xanderdunn** [comment]
That's great to hear, thanks! Happy to test the fix anytime

**Comment by xanderdunn** [comment]
This is fixed in neuronx-cc 2.8 beta:
```
$ neuronx-cc --version
NeuronX Compiler version 2.8.0.13+a76a82727

Running on AMI ami-060e1cfd2b2588d6b
Running in region usw2-az4
Running on python 3.8.10

HWM version 2.8.0.3-2b7c6da39
NumPy version 1.21.6
```

Thanks so much, we got our first successful run of a sizable transformer model across 4xtrn1n.32xlarge instances with this fix! 🥳
